### PR TITLE
Printing and testing for EntityTypeManager

### PR DESCRIPTION
**Pull request by witchel**
This PR is mostly in support of having graph update use the EntityTypeManager (enterprise PR #2049).  It does the following.

- Adds debugging routine to print the types within a manager.
- Adds a rudimentary test that also acts as documentation for how the type manager works.

### CODE REVIEW

**Review by amberhassaan — CHANGES_REQUESTED** [review]
Correct me if I'm wrong, but your implementation of `valid` bit is invalid (as in buggy). 
In order to do it right, you'd have to define move constructor and assignment operator that mark the moved object invalid, because right now, nothing marks the moved manager invalid. 

This is also why I argued against the use of valid bit, and using an `is_valid()` implementation that checks the member maps for being empty because they become empty after having been moved. 

I'll take a look at the enterprise PR some time tomorrow.

File: libsupport/include/katana/EntityTypeManager.h
Comment (amberhassaan) [inline libsupport/include/katana/EntityTypeManager.h]: This can be expensive if we are using the hash function on very large maps/sets, e.g. O(V) or O(E). I'd like to encourage an API that works with type DI sets (`std::bitset`) instead of type name sets (`set<string>`)

**Comment by witchel** [comment]
> Correct me if I'm wrong, but your implementation of `valid` bit is invalid (as in buggy).
> In order to do it right, you'd have to define move constructor and assignment operator that mark the moved object invalid, because right now, nothing marks the moved manager invalid.
> 
> This is also why I argued against the use of valid bit, and using an `is_valid()` implementation that checks the member maps for being empty because they become empty after having been moved.
> 
> I'll take a look at the enterprise PR some time tomorrow.

Arrghh.  I even tested it, but the crash I got was from my calling `value()` on a error result.

I thought the move does  a swap on the members which leaves them with their default values.  But I was wrong.

I tried the empty map approach, but as I wrote in slack, there is an initialization path that calls several public functions.  These functions cannot assert that the object is valid because for most of initialization the map is empty.  Splitting these functions into a private part that does not check validity and a public one that does became unwieldy. 

I will likely just roll this feature back.

File: libsupport/include/katana/EntityTypeManager.h
Comment (witchel) [reply libsupport/include/katana/EntityTypeManager.h]: Type name sets are small: order of number of properties per node (10s) or edge (1 in Cypher).

I need a set of strings interface because that is what I get from the Cypher engine.  If you or @roshandathathri tell them to pass me bitsets then I can recode my public API to only accept bitsets.  But I think that might have problems with bitsets going stale as label sets change.  The last time we had this discussion, Roshan called the string set the source of truth.

File: libsupport/include/katana/EntityTypeManager.h
Comment (scober) [inline libsupport/include/katana/EntityTypeManager.h]: nit: I think there has been some discussion about what this string should be (or if there should be a string here at all). I would like to propose "Unknown Type". I think it is more descriptive and I don't think we need to restrict ourselves to CamelCase with these names.

File: libsupport/include/katana/EntityTypeManager.h
Comment (scober) [reply libsupport/include/katana/EntityTypeManager.h]: +1 for `std::set<std::string>` as the source of truth/canonical form.

File: libsupport/include/katana/EntityTypeManager.h
Comment (scober) [inline libsupport/include/katana/EntityTypeManager.h]: nit: I find this comment a little confusing. Why would the set be any kind of name?

File: libsupport/include/katana/EntityTypeManager.h
Comment (scober) [inline libsupport/include/katana/EntityTypeManager.h]: Should we put this string in a `constexpr` at the top of the file?

File: libsupport/include/katana/EntityTypeManager.h
Comment (scober) [inline libsupport/include/katana/EntityTypeManager.h:452]: This is awesome

File: libsupport/test/type-manager.cpp
Comment (scober) [inline libsupport/test/type-manager.cpp]: `num_entities` isn't used again in this function

**Review by scober — APPROVED** [review]
LGTM

File: libsupport/include/katana/EntityTypeManager.h
Comment (roshandathathri) [reply libsupport/include/katana/EntityTypeManager.h]: Yes, the source of truth is a set of strings.

Comment (roshandathathri) [reply libsupport/include/katana/EntityTypeManager.h]: Is this needed for `std::set`? I thought not. It's needed for `std::unordered_set` or `std::unordered_hash`. Can we avoid using those?

File: libsupport/include/katana/EntityTypeManager.h
Comment (roshandathathri) [reply libsupport/include/katana/EntityTypeManager.h]: This is valid after #577. It should return `kUnknownEntityType`.

File: libsupport/include/katana/EntityTypeManager.h
Comment (roshandathathri) [inline libsupport/include/katana/EntityTypeManager.h]: nit: `PrintEntityTypes`

File: libsupport/test/type-manager.cpp
Comment (roshandathathri) [inline libsupport/test/type-manager.cpp]: This will be empty after #577.

File: libsupport/test/type-manager.cpp
Comment (roshandathathri) [inline libsupport/test/type-manager.cpp]: Not true after #577.

File: python/test/test_entity_type_manager.py
Comment (roshandathathri) [inline python/test/test_entity_type_manager.py]: Thanks. Didn't know I had to change this as well in #577.

**Review by roshandathathri — APPROVED** [review]
LGTM.

It's fine if this merges before #577 but please address the struct hash question before merging.

File: libsupport/include/katana/EntityTypeManager.h
Comment (witchel) [reply libsupport/include/katana/EntityTypeManager.h:452]: It is the little things in life.

File: libsupport/include/katana/EntityTypeManager.h
Comment (witchel) [reply libsupport/include/katana/EntityTypeManager.h]: Removed.

File: libsupport/test/type-manager.cpp
Comment (witchel) [reply libsupport/test/type-manager.cpp]: Yup, I fixed it.

File: libsupport/test/type-manager.cpp
Comment (witchel) [reply libsupport/test/type-manager.cpp]: Fixed

**Comment by witchel** [comment]
> LGTM.
> 
> It's fine if this merges before #577 but please address the struct hash question before merging.

Thanks for keeping in mind the bigger picture!  The hash function is no longer used and I eliminated it.

**Comment by witchel** [comment]
> Correct me if I'm wrong, but your implementation of `valid` bit is invalid (as in buggy). In order to do it right, you'd have to define move constructor and assignment operator that mark the moved object invalid, because right now, nothing marks the moved manager invalid.
> 
> This is also why I argued against the use of valid bit, and using an `is_valid()` implementation that checks the member maps for being empty because they become empty after having been moved.
> 
> I'll take a look at the enterprise PR some time tomorrow.

Amber, are you still intending to block this PR?  I got rid of the IsValid code.

**Comment by witchel** [comment]
> > Correct me if I'm wrong, but your implementation of `valid` bit is invalid (as in buggy). In order to do it right, you'd have to define move constructor and assignment operator that mark the moved object invalid, because right now, nothing marks the moved manager invalid.
> > This is also why I argued against the use of valid bit, and using an `is_valid()` implementation that checks the member maps for being empty because they become empty after having been moved.
> > I'll take a look at the enterprise PR some time tomorrow.
> 
> Amber, are you still intending to block this PR? I got rid of the IsValid code.

@amberhassaan, this PR is blocked on your request.

**Comment by amberhassaan** [comment]
> > > Correct me if I'm wrong, but your implementation of `valid` bit is invalid (as in buggy). In order to do it right, you'd have to define move constructor and assignment operator that mark the moved object invalid, because right now, nothing marks the moved manager invalid.
> > > This is also why I argued against the use of valid bit, and using an `is_valid()` implementation that checks the member maps for being empty because they become empty after having been moved.
> > > I'll take a look at the enterprise PR some time tomorrow.
> > 
> > 
> > Amber, are you still intending to block this PR? I got rid of the IsValid code.
> 
> @amberhassaan, this PR is blocked on your request.

Sorry for ignoring this (and your reminders). Please do ping me (everywhere) if my review is needed.

**Review by amberhassaan — APPROVED** [review]

**Pull request merged**
